feat: expand `.ddev/.env.*` files on `ddev start`, add `ddev dotenv`, add variable expansion to `ddev add-on get`, fixes #4430 (#6406)
ddev/ddev: docs/content/users/extend/additional-services.md

## Adding Custom Configuration to an Add-on

> Sometimes it's necessary to add custom configuration to an add-on. For example, in [`ddev-redis`](https://github.com/ddev/ddev-redis) the `image` is set to `image: redis:${REDIS_VERSION:-6-bullseye}`. If you wanted to change this to use `7.0-alpine` instead, you would have three choices:

Reviewer: This is a nit, but I would recommend not suggesting a completely different OS and build, as they won't necessarily be compatible.

```suggestion
Sometimes it's necessary to add custom configuration to an add-on. For example, in [`ddev-redis`](https://github.com/ddev/ddev-redis) the `image` is set to `image: redis:${REDIS_VERSION:-6-bullseye}`. If you wanted to change this to use `7.0-bullseye` instead, you would have three choices:
```

> 1. With DDEV v1.23.4+, the override can be done using a special `.ddev/.env.*` file, where `*` is the name of the add-on, in this case `.ddev/.env.redis`:

Author: ```suggestion
1. With DDEV v1.23.5+, the override can be done using a special `.ddev/.env.*` file, where `*` is the name of the add-on, in this case `.ddev/.env.redis`:
```

    ```bash
    ddev get ddev/ddev-redis --environment="REDIS_VERSION=7.0-alpine"
    ddev restart
    ```

> 2. Remove the `#ddev-generated` line in `.ddev/docker-compose.redis.yaml` and edit the relevant line.

Reviewer: Why don't we just stop recommending this, since we have better ways? The general docs explain this a few places, but it's not even needed here.

```suggestion
```

> 3. Add a `.ddev/docker-compose.redis_extra.yaml` with the contents:

Author: ```suggestion
2. Add a `.ddev/docker-compose.redis_extra.yaml` with the contents:
```

    ```yaml
    services:
      redis:

>         image: redis:7.0-alpine

Reviewer: Recommend using similar example, 
```suggestion
        image: redis:7.0-bullseye
```

    ```

> Using the first (`.env.redis`) and the third option (`docker-compose.redis_extra.yaml`) allows you to update to future versions of `ddev-redis` without losing your configuration, and without the upgrade being blocked because you removed the `#ddev-generated` line from the upstream `docker-compose.redis.yaml`.

Reviewer: ```suggestion
Using the options here allow you to update to future versions of `ddev-redis` without losing your configuration, and without the upgrade being blocked because you removed the `#ddev-generated` line from the upstream `docker-compose.redis.yaml`.
```

## Adding Custom Configuration to an Add-on

Sometimes it's necessary to add custom configuration to an add-on. For example, in [`ddev-redis`](https://github.com/ddev/ddev-redis) the `image` is set to `image: redis:${DDEV_REDIS_VERSION:-6-bullseye}`. If you wanted to change this to use `7.0-bookworm` instead, you would have two choices:

> 1. With DDEV v1.23.5+, the override can be done using a special `.ddev/.env.*` file, where `*` is the name of the add-on (and its service), in this case `.ddev/.env.redis`:

Author: `something` in `.env.something` is only relevant if env variables are to be passed inside a service.

Otherwise the filename is irrelevant, it will be expanded by `docker-compose config` anyway.
I use the add-on name to name this `.env.*` file somehow.

    ```bash
    ddev add-on get ddev/ddev-redis --environment="DDEV_REDIS_VERSION=7.0-alpine"
    ddev restart
    ```

    You can override `.ddev/.env.*` with `.ddev/.env.*.override` file in case you need more customization.

>     !!!warning "Environment variables cannot be overridden in some cases"

Author: I should rethink how it works. Because I don't like it.

        If your `docker-compose.*.yaml` file already has some variables in the `environment` or `env_file` stanza, you cannot override them with `.ddev/.env.*` or `.ddev/.env.*.override` files.
        

>         If this is the case, move these variables to the appropriate `.ddev/.env.*` file to be able to override them with a `.ddev/.env.*.override` file.

Author: This is why I removed this section from the app template:

```yaml
    {{ if .EnvFile }}
    env_file: '{{ .EnvFile }}'
    {{ end }}
```

Otherwise, I couldn't override any existing variables from `.ddev/.env` with `.ddev/.env.override`.

2. Add a `.ddev/docker-compose.redis_extra.yaml` with the contents:

    ```yaml
    services:
      redis:
        image: redis:7.0-bookworm
    ```

> Using the options here allow you to update to future versions of `ddev-redis` without losing your configuration, and without the upgrade being blocked because you removed the `#ddev-generated` line from the upstream `docker-compose.redis.yaml`.

Author: I added a note below.

## Adding Custom Configuration to an Add-on

Sometimes it's necessary to add custom configuration to an add-on. For example, in [`ddev-redis`](https://github.com/ddev/ddev-redis) the `image` is set to `image: redis:${REDIS_TAG:-6-bullseye}` (it is important to have the default value for `$REDIS_TAG`). If you wanted to change this to use `7-bookworm` instead, you would have two choices:

> 1. With DDEV v1.23.5+, the override can be done using a special `.ddev/.env.*` file, where `*` is the name of the add-on, in this case `.ddev/.env.redis` (file is created automatically):

Reviewer: I think we have a little bit of trouble here. 

The name of an add-on *may* be the same as the name of the *service*. But it may not be. There's nothing that forces that. In fact, in ddev-ioncube, the name is nonstandard... and there is also not a docker-compose.*.yaml, so it's impossible to do a .env anyway. 

But if there *were* a `docker-compose.*.yaml`, it might contain more than one service. 

I think what we probably have here at this point is just a naming problem. I think you're completely handling all *services* by looking and the fully constructed compose yaml. But that may not be the exact name of the add-on. And if a `docker-compose.*.yaml` had more than one service in it (I think some do) then this would mostly work OK?

> 1. With DDEV v1.23.5+, the override can be done using a special `.ddev/.env.*` file, where `*` is the name of the add-on, in this case `.ddev/.env.redis` (file is created automatically):

Author: The add-on name matters only in the `ddev add-on get --something=value` command, and as you mention, this is a problem.

And luckily we can make it better and obvious using a new command `ddev dotenv .env.ioncube --flags`, that I described in:
- https://github.com/ddev/ddev/pull/6406#issuecomment-2363869271

The `install.yaml` is a simple YAML file with a few main sections:

* `name`: The add-on name. This will be used in the various `ddev add-on` commands. If the repository name is `ddev-solr`, for example, a good `name` will be `solr`.

> * `pre_install_actions`: an array of Bash statements or scripts to be executed before `project_files` are installed. The actions are executed in the context of the target project’s root directory. With DDEV v1.23.5+, variables from the `.ddev/.env.name` file, where `name` is the name of the add-on, are available in this context.

Reviewer: ```suggestion
* `pre_install_actions`: an array of Bash statements or scripts to be executed before `project_files` are installed. The actions are executed in the context of the target project’s root directory. With DDEV v1.23.5+, variables from the `.ddev/.env.name` file, where `name` is the name of the service in the add-on's `docker-compose.*.yaml`, are available in this context.
```

Author: And here `.env.name` is always the name of the add-on. I can't use the service name here because add-ons may not have services, and I can't get the service name without running `docker-compose config`, which creates a catch-22.

    ```

* `dependencies`: an array of add-ons that this add-on depends on.

> * `post_install_actions`: an array of Bash statements or scripts to be executed after `project_files` and `global_files` are installed. The actions are executed in the context of the target project’s `.ddev` directory. With DDEV v1.23.5+, variables from the `.ddev/.env.name` file, where `name` is the name of the add-on, are available in this context.

Reviewer: So here, the .env.name is for the add-on name, true? But elsewhere, service name?

Author: Yes.
